_studio/shared/umc/codec/av1_dec/src/umc_av1_frame.cpp

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2012-2020 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "umc_defs.h"
+#ifdef MFX_ENABLE_AV1_VIDEO_DECODE
+
+#include "umc_av1_dec_defs.h"
+#include "umc_av1_frame.h"
+#include "umc_av1_utils.h"
+#include <limits>
+#include <algorithm>
+
+#include "umc_frame_data.h"
+#include "umc_media_data.h"
+
+namespace UMC_AV1_DECODER
+{
+    TileSet::TileSet(UMC::MediaData* in, TileLayout const& layout_)
+        : layout(layout_)
+    {
+        if (!in)
+            throw av1_exception(UMC::UMC_ERR_NULL_PTR);
+
+        TileLocation const& lastTile = layout.back();
+        size_t const size = lastTile.offset + lastTile.size;
+        source.Alloc(size);
+        if (source.GetBufferSize() < size)
+            throw av1_exception(UMC::UMC_ERR_ALLOC);
+
+        const uint8_t *src = reinterpret_cast <const uint8_t*> (in->GetDataPointer());
+        uint8_t *dst = reinterpret_cast <uint8_t*> (source.GetDataPointer());
+        std::copy(src, src + size, dst);
+
+        source.SetDataSize(size);
+    }
+
+    size_t TileSet::Submit(uint8_t* bsBuffer, size_t spaceInBuffer, size_t offsetInBuffer, TileLayout& layoutWithOffset)
+    {
+        if (submitted)
+            return 0;
+
+        if (!bsBuffer)
+            throw av1_exception(UMC::UMC_ERR_NULL_PTR);
+
+        uint8_t* data = static_cast<uint8_t*>(source.GetDataPointer());
+        const size_t length = std::min<size_t>(source.GetDataSize(), spaceInBuffer);
+
+        MFX_INTERNAL_CPY(bsBuffer + offsetInBuffer, data, length);
```

Comment:
https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/Coding_guidelines.md#memory-management

Reply:
MFX_INTERNAL_CPY will use std::copy which is defined at the following line. Need the other change?
https://github.com/Intel-Media-SDK/MediaSDK/blob/master/_studio/shared/umc/core/umc/include/umc_defs.h#L56

Reply:
Thanks for checking! Addressed